runtime/jvmti/jvmtiThread.c

```diff
@@ -528,32 +528,39 @@ jvmtiGetThreadInfo(jvmtiEnv *env,
 						jobject tempGroup = NULL;
 
 						vm->internalVMFunctions->internalExitVMToJNI(currentThread);
-						if ((NULL == constants) || (NULL == id)) {
-							/* Cache class and field id. */
+						if (NULL == constants) {
+							/* Cache the class. */
 							jclass localRef = (*jniEnv)->FindClass(jniEnv, "java/lang/Thread$Constants");
 							Assert_JVMTI_notNull(localRef);
 							constants = (jclass)((*jniEnv)->NewGlobalRef(jniEnv, localRef));
-							if (NULL == constants) {
-								releaseVMThread(currentThread, targetThread, thread);
-								j9mem_free_memory(name);
+							if (NULL != constants) {
+								vm->jlThreadConstants = constants;
+							} else {
 								rc = JVMTI_ERROR_OUT_OF_MEMORY;
-								goto exit;
 							}
+						}
+
+						if ((NULL != constants) && (NULL == id)) {
+							/* Cache the field id. */
 							id = (*jniEnv)->GetStaticFieldID(jniEnv, constants, "VTHREAD_GROUP", "Ljava/lang/ThreadGroup;");
-							if (NULL == id) {
-								releaseVMThread(currentThread, targetThread, thread);
-								j9mem_free_memory(name);
+							if (NULL != id) {
+								vm->vthreadGroupID = id;
+							} else {
 								rc = JVMTI_ERROR_OUT_OF_MEMORY;
-								goto exit;
 							}
+						}
 
-							vm->jlThreadConstants = constants;
-							vm->vthreadGroupID = id;
+						if ((NULL != constants) && (NULL != id)) {
+							tempGroup = (*jniEnv)->GetStaticObjectField(jniEnv, constants, id);
+							Assert_JVMTI_notNull(tempGroup);
 						}
 
-						tempGroup = (*jniEnv)->GetStaticObjectField(jniEnv, constants, id);
-						Assert_JVMTI_notNull(tempGroup);
 						vm->internalVMFunctions->internalEnterVMFromJNI(currentThread);
+						/* VM access is required before invoking releaseVMThread. */
```

Comment:
I think the comment belongs before the call to enter the VM.

Reply:
Updated.